actix-session/src/storage/sqlite.rs

```diff
@@ -0,0 +1,348 @@
+///   let store = SqliteSessionStore::new(pool, true).unwrap();
+///
+///   HttpServer::new(move || {
+///       App::new()
+///           .wrap(SessionMiddleware::new(
+///                 store.clone(),
+///                 secret_key.clone()
+///            ))
+///           .default_service(web::to(|| HttpResponse::Ok()))
+///   })
+///   .bind(("127.0.0.1", 8080))?
+///   .run()
+///   .await
+/// }
+/// ```
+///
+/// # Implementation notes
+/// `SqliteSessionStore` uses rusqlite, r2d2 and r2d2_sqlite.
+///
+/// [`rusqlite`]: https://github.com/rusqlite/rusqlite
+/// [`r2d2`]: https://github.com/sfackler/r2d2
+/// [`r2d2_sqlite`]: https://github.com/ivanceras/r2d2-sqlite
+#[cfg_attr(docsrs, doc(cfg(feature = "sqlite-session")))]
+#[derive(Clone)]
+pub struct SqliteSessionStore {
+    configuration: CacheConfiguration,
+    pool: Pool<SqliteConnectionManager>,
+}
+
+#[derive(Clone)]
+struct CacheConfiguration {
+    cache_keygen: Arc<dyn Fn(&str) -> String + Send + Sync>,
+}
+
+impl Default for CacheConfiguration {
+    fn default() -> Self {
+        Self {
+            cache_keygen: Arc::new(str::to_owned),
+        }
+    }
+}
+
+impl SqliteSessionStore {
+    /// Create a new instance of [`SqliteSessionStore`] using the default configuration.
+    /// It takes two required inputs to create a new instance of [`SqliteSessionStore`] - a
+    /// pool of connections [`Pool<SqliteConnectionManager>`] and a boolean which specifies
+    /// whether default garbage connection is enabled/disabled.
+    ///
+    /// Default garbage collection of stale sessions happens through a database trigger, which
+    /// executes before every insert on the `sessions` table.
+    pub fn new(
+        pool: Pool<SqliteConnectionManager>,
+        garbage_collect: bool,
+    ) -> Result<Self, anyhow::Error> {
+        // create sessions table if it doesn't exist
+        let conn = pool.get().unwrap();
+        conn.execute(
+            "CREATE TABLE IF NOT EXISTS sessions (
+          id INTEGER NOT NULL,
+          session_key	TEXT NOT NULL,
+          session_state	TEXT NOT NULL,
+          expiry INTEGER NOT NULL,
+          PRIMARY KEY(id AUTOINCREMENT)
+        )",
+            [],
+        )
+        .map_err(Into::into)
+        .map_err(LoadError::Other)?;
+
+        // in order to garbage collect stale sessions, we will use a trigger
+        if garbage_collect {
+            conn.execute(
+                "
+            CREATE TRIGGER IF NOT EXISTS garbage_collect 
+              BEFORE INSERT 
+              ON sessions
+            BEGIN
+              DELETE FROM sessions WHERE expiry < STRFTIME('%s');
+            END;
+          ",
+                [],
+            )
+            .map_err(Into::into)
+            .map_err(LoadError::Other)?;
+        } else {
+            conn.execute("DROP TRIGGER IF EXISTS garbage_collect", [])
+                .map_err(Into::into)
+                .map_err(LoadError::Other)?;
+        }
```

Comment:
This is a one-off setup action that should not be executed every single time somebody tries to _use_ Sqlite as a session backend in their applications.
It should live in a separate routine, `new` should assume that the initialization has already taken place.